### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by kulich-d**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,7 @@
-# master-dissertation
\ No newline at end of file
+# master-dissertation
+
+example of usage:
```
Comment (mishuk-sk) [inline README.md]:
```suggestion
example usage:
```

File: README.md [hunk 2]
```diff
@@ -1 +1,7 @@
...
+
+     python3 main.py --video_path /Users/diana.kulich/Documents/Masters/dissertation/data/каражан/ch01_20210909192521.mp4 --save_path /Users/diana.kulich/Documents/Masters/dissertation/exp --exp_name "exp_name"
```
Comment (mishuk-sk) [inline README.md]:
```suggestion
     python3 main.py --video_path <video_path> --save_path <path_to_save> --exp_name <exp_name>
```

File: detectors_comparison.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import os
+
+import cv2
+import numpy as np
+
+video_name = "ch01_20210909192521"
+path_mediapipe = "/Users/diana.kulich/Documents/Masters/dissertation/mediapipe_results"
+path_openpose = "/Users/diana.kulich/Documents/Masters/dissertation/open_pose_results"
+path_yolo = "/Users/diana.kulich/Documents/Masters/dissertation/yolov7/runs/test"
+save_path = "/Users/diana.kulich/Documents/Masters/dissertation/detectors_comparison"
```
Comment (mishuk-sk) [inline detectors_comparison.py]: Shouldn't  it be dynamic paths? Like shouldn't they be provided from comand line?

File: detectors_comparison.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    image_yolo = cv2.resize(image_yolo, (960, 480))
+    image_yolo = cv2.putText(image_yolo, 'Yolo', (50, 50), cv2.FONT_HERSHEY_SIMPLEX,
+                                 1, (255, 0, 0), 1, cv2.LINE_AA)
+
+    f_name = os.path.join(save_path, video_name, "frame%d.jpg" % count)
+    big_image = np.stack((image_mediapipe, image_openpose, image_yolo)).reshape((480*3, 960, 3))
+    cv2.imwrite(f_name, big_image)  # save frame as JPEG file
+    print(count)
+
+    count += 1
```
Comment (mishuk-sk) [inline detectors_comparison.py]: What is the point of this forever loop?

File: main.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import os
+
+import cv2
+
+import pipeline
+from video_rider import VideoReader
+
+
+def treadmill_coordinates(image):
+    camera1_coord_template = [(300 / 960, 0 / 480), (750 / 960, 450 / 480)]
```
Comment (mishuk-sk) [inline main.py:11]: Should it be hardcoded?

File: main.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    image = cv2.rectangle(image, camera_coord[0], camera_coord[1], (0, 0, 0), 3)  # 1 camera
+    return image
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--video_path', type=str, required=True)
+    parser.add_argument('--save_path', type=str, required=True)
+    parser.add_argument('--exp_name', type=str, required=True)
+    parser.add_argument('--visualize', type=bool, default=True)
```
Comment (mishuk-sk) [inline main.py:25]: would be nice to have help tips in the parser like:
```python
parser.add_argument('--video_path', type=str, required=True, description='What the program does')
```

File: main.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    args = parser.parse_args()
+    video_name: str = args.video_path.split("/")[-1].split(".")[0]
+
+    args.save_path = os.path.join(args.save_path, args.exp_name + video_name)
+    if not os.path.exists(args.save_path):
+        os.mkdir(args.save_path)
+    video = VideoReader(args.video_path)
+    pipeline.main(video, args.save_path, args.visualize)
```
Comment (mishuk-sk) [inline main.py]: usually there is only one main function as an entry point to application, maybe consider renaming to `run`?

File: posprocessing.py
Comment (mishuk-sk) [inline posprocessing.py:1]: `pose`/`post`/`position`?

File: posprocessing.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+def frames_borders(data, minf, maxf):
+    filter_data = []
+    for f in data:
+        if f > minf and f < maxf:
+            filter_data.append(f)
+    return filter_data
+
+
+def skip_unnecessary_frames(frames_with_steps, frames_start_second_double_support, frames_first_double_support,
```
Comment (mishuk-sk) [inline posprocessing.py]: would it make sense to have funciton signature as 
```python
def skip_unnecessary_frames(frames_with_steps, frames _to_filter)
``` 
and call it 3 times for each of `frames_start_second_double_support`, `frames_first_double_support`, `frames_second_double_support` as a parameter?

File: read_russian_json.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import json
+import os
+
+import numpy as np
+
+save_path = "/Users/diana.kulich/Documents/Masters/dissertation/exp/Karjan_girl_all_processing_new_anglesch01_20210909192521"
```
Comment (mishuk-sk) [inline read_russian_json.py]: shouldn't it be dynamic?

File: read_russian_json.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+with open(os.path.join(save_path, "all_information.json"), encoding='ascii') as f:
+    data = json.load(f)
+print(data)
+for k, v in data.items():
+    v = [float(i) for i in v]
+    print(f"{k}: median {round(np.median(v), 2)}, max {round(np.max(v), 2)},  min {round(np.min(v), 2)},  ")
+    print( v)
```
Comment (mishuk-sk) [inline read_russian_json.py]: wrap in function?

File: utils.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    angle_barycenter = dtw_barycenter_averaging_subgradient(X, max_iter=50, tol=1e-3)
+    return np.squeeze(angle_barycenter)
+
+
+def save_report(save_path, report_information):
+    with open(os.path.join(save_path, "report_information.json"), "w", encoding='ascii') as json_file:
+        json.dump(report_information, json_file, default=str)
+    with open(os.path.join(save_path, "report_information.json"), encoding='ascii') as f:
+        data = json.load(f)
+    print(data)
```
Comment (mishuk-sk) [inline utils.py:103]: EOF (missing new line at the and of file)

Comment (mishuk-sk) [reply utils.py:103]: Not important

File: step_segmentation.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        if path_dtw_1[p] in peaks_y_2:
+            all_picks.append(p)
+            all_picks_mean.append(np.mean([p, path_dtw_1[p]]))
+    for p in peaks_y_2:
+        if path_dtw_2[p] in peaks_y:
+            all_picks.append(p)
+            all_picks_mean.append(np.mean([p, path_dtw_2[p]]))
+
+    print(f"common_picks: {all_picks}")
+    print(f"common_mean_picks: {all_picks_mean}")  # это и есть дабл сапорт первый
```
Comment (mishuk-sk) [inline step_segmentation.py]: первый? Функция называется второй

File: video_rider.py
Comment (mishuk-sk) [inline video_rider.py:1]: 💪

File: visualization.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+    visualize_landmarks_coordinates(skeleton)
+    annotated_image = concat_visulisations(annotated_image)
+    return annotated_image
+
+
+def concat_visulisations(image):
+    graph_img_x = cv2.resize(cv2.imread("temp_x.png"), (image.shape[1], image.shape[0]))
+    graph_img_y = cv2.resize(cv2.imread("temp_y.png"), (image.shape[1], image.shape[0]))
+    graph_img_z = cv2.resize(cv2.imread("temp_z.png"), (image.shape[1], image.shape[0]))
```
Comment (mishuk-sk) [inline visualization.py]: Where do these names come from?

Comment (mishuk-sk) [reply visualization.py]: file names

File: visualization.py
Comment (mishuk-sk) [inline visualization.py:1]: 💪

**Review by mishuk-sk — APPROVED** [review]
I have not really wen't through the logic of the code only highlighed the things popping out. 

Will have a deeper look later 😄

File: main.py (hunk 5 above)
Comment (kulich-d) [reply main.py:11]: yes

File: visualization.py (hunk 13 above)
Comment (kulich-d) [reply visualization.py]: I need them for create visualization
They were created by this pipeline, and removed when pipeline finished

File: step_segmentation.py (hunk 12 above)
Comment (kulich-d) [reply step_segmentation.py]: Первый)

Comment (kulich-d) [reply step_segmentation.py]: Ой, второй, короче, функция называется правильно

File: visualization.py (hunk 13 above)
Comment (mishuk-sk) [reply visualization.py]: It would be nice to store them as constants if they are used elsewhere

File: data_filtering.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
+from scipy.signal import savgol_filter
+
+
+def filter_data_coordinates(data):
```
Comment (mishuk-sk) [inline data_filtering.py:5]: I don't know how to understand what this function acually expects in `data`... Is it possible to define it somehow for python? (and is it needed? Hard for me to judge)

File: data_filtering.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    y_filter = np.array(data)[:, 1]
+    y_filter = savgol_filter(y_filter, window_size, min(window_size - 1, 2), mode='interp')
+    return x_filter, y_filter
+
+
+def filter_data_angle(data):
+    x_filter = np.array(data)
+    window_size = min(len(x_filter), 15)
+    window_size = window_size if window_size % 2 == 1 else window_size - 1
+    x_filter = savgol_filter(x_filter, window_size, min(window_size - 1, 2), mode='interp')
```
Comment (mishuk-sk) [inline data_filtering.py:20]: aren't these functions the same?  can't you just define filter outside of function and call `filter_data_angle` twice to get the same data `fileter_data_coordinates` returns? If so, maybe consider renaming to `apply_savgol_filter(data, filter)` signature?

File: get_characteristics.py [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    knee_ankle = np.abs([knee[0] - ankle[0], knee[1] - ankle[1]])
+    knee_heel = np.abs([knee[0] - heel[0], knee[1] - heel[1]])
+    ankle_heel = np.abs([heel[0] - ankle[0], heel[1] - ankle[1]])
+
+    knee_ankle_len = math.sqrt((knee_ankle[0] * video.width) ** 2 + (knee_ankle[1] * video.height) ** 2)
+    knee_heel_len = math.sqrt((knee_heel[0] * video.width) ** 2 + (knee_heel[1] * video.height) ** 2)
+    ankle_heel_len = math.sqrt((ankle_heel[0] * video.width) ** 2 + (ankle_heel[1] * video.height) ** 2)
+
+    return math.degrees(math.acos(
+        (knee_ankle_len ** 2 + ankle_heel_len ** 2 - 2 * knee_ankle_len * ankle_heel_len) / knee_heel_len ** 2))
```
Comment (mishuk-sk) [inline get_characteristics.py:51]: Seems like this 3 functions also do the same job, while just operating on different datapoints (not sure though, hard to tell without context). Feels like this functions only calculate some angles between smth. Why not to unify to single function?

File: main.py (hunk 5 above)
Comment (mishuk-sk) [reply main.py:11]: can't resolve, resolve please (no permissions most likely)

File: pipeline.py [hunk 17]
```diff
@@ -0,0 +1,65 @@
+import math
+import os
+from dataclasses import dataclass
+
```
Comment (mishuk-sk) [inline pipeline.py:4]: why empty line in imports? Is it python codestyle-related?

File: posprocessing.py
Comment (mishuk-sk) [reply posprocessing.py:1]: file name is unclear

File: report.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+            report_info["Тазобедренный сустав период переноса вертикальное положение большеберцовой кости"].append(
+                get_characteristics.angle_hip_joint(data_report, statistic_frame, video))
+            report_info["Коленный сустав период переноса вертикальное положение большеберцовой кости"].append(
+                get_characteristics.angle_knee_joint(data_report, statistic_frame, video))
+            report_info["Голеностопный сустав период переноса вертикальное положение большеберцовой кости"].append(
+                get_characteristics.angle_ankle_joint(data_report, statistic_frame, video))
+
+    if visualize:
+        visualization.report_visualization(state_dict, step_start, save_path)
+    utils.save_report(save_path, report_info)
```
Comment (mishuk-sk) [inline report.py:120]: Doesn't this cycles do the same? Maybe consider extracting to function?

File: step_segmentation.py [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+            all_picks.append(p)
+            all_picks_mean.append(np.mean([p, path_dtw_1[p]]))
+    for p in peaks_y_2:
+        if path_dtw_2[p] in peaks_y:
+            all_picks.append(p)
+            all_picks_mean.append(np.mean([p, path_dtw_2[p]]))
+
+    print(f"common_picks: {all_picks}")
+    print(f"common_mean_picks: {all_picks_mean}")
+    return all_picks_mean
```
Comment (mishuk-sk) [inline step_segmentation.py:74]: Does the only differ with `"right"` and `"left"` string literals? maybe could be the same funciton accepting `"right"` or `"left"`?

**Review by mishuk-sk — COMMENTED** [review]
Done with review as much as I could review python code 😄

**Pull request merged**